rust/routee-compass-powertrain/src/routee/energy_traversal_model.rs

```diff
@@ -96,44 +83,53 @@ impl TraversalModel for EnergyTraversalModel {
             get_vehicle_state_from_state(state),
         )?;
 
-        let mut energy = energy_result.energy;
-
-        // for now we need to truncate the energy at zero until we can handle these being negative
-        if energy.as_f64() < 0.0 {
-            energy = Energy::new(ZERO_ENERGY);
-            log::debug!("negative energy encountered, setting to 1e-9");
-        }
-
-        let total_cost = create_cost(energy, time, self.energy_cost_coefficient);
-
         let updated_state = update_state(state, distance, time, energy_result.updated_state);
-        let result = TraversalResult {
-            total_cost,
-            updated_state,
-        };
-        Ok(result)
+        Ok(updated_state)
     }
 
-    fn serialize_state(&self, state: &TraversalState) -> serde_json::Value {
-        let distance = get_distance_from_state(state);
-        let time = get_time_from_state(state);
-        let vehicle_state = get_vehicle_state_from_state(state);
-        let vehicle_state_summary = self.vehicle.serialize_state(vehicle_state);
-        serde_json::json!({
-            "distance": distance,
-            "time": time,
-            "vehicle": vehicle_state_summary,
-        })
+    fn access_edge(
+        &self,
+        _v1: &Vertex,
+        _src: &Edge,
+        _v2: &Vertex,
+        _dst: &Edge,
+        _v3: &Vertex,
+        _state: &TraversalState,
+    ) -> Result<Option<TraversalState>, TraversalModelError> {
+        Ok(None)
     }
 
-    fn serialize_state_info(&self, state: &TraversalState) -> serde_json::Value {
-        let vehicle_state = get_vehicle_state_from_state(state);
-        let vehicle_state_info = self.vehicle.serialize_state_info(vehicle_state);
-        serde_json::json!({
-            "distance_unit": self.service.output_distance_unit,
-            "time_unit": self.service.output_time_unit,
-            "vehicle_info": vehicle_state_info,
-        })
+    fn estimate_traversal(
+        &self,
+        src: &Vertex,
+        dst: &Vertex,
+        state: &TraversalState,
+    ) -> Result<TraversalState, TraversalModelError> {
+        let distance = haversine::coord_distance(
+            src.coordinate,
+            dst.coordinate,
+            self.service.output_distance_unit,
+        )
+        .map_err(TraversalModelError::NumericError)?;
+
+        if distance == Distance::ZERO {
+            return Ok(state.clone());
+        }
+
+        let time: Time = Time::create(
+            self.service.max_speed,
+            self.service.speeds_table_speed_unit,
+            distance,
+            self.service.output_distance_unit,
+            self.service.output_time_unit.clone(),
+        )?;
+
+        let best_case_result = self
+            .vehicle
+            .best_case_energy_state((distance, self.service.output_distance_unit), state)?;
```

Comment:
OH! I think this state just needs to be wrapped in `get_vehicle_state_from_state(state)` to extract the vehicle specific subset of the whole traversal state. 
```suggestion
            .best_case_energy_state((distance, self.service.output_distance_unit), get_vehicle_state_from_state(state))?;
```

Reply:
nice catch! after fixing this, the energy-optimal search runtime is down for me from 40 seconds to 0 seconds. so, now, wondering why that went so fast, and why time and distance are taking so long in comparison. updated table:

name | parameters | runtime | tree size | route distance | route time | route energy | total cost |
--- | --- | --- | --- | --- | --- | --- | --- |
distance-optimal | distance = 1 | 2.5 sec | 61044 | 17.60 miles | 42.25 minutes | 4.68kWh | 64.54 |
time-optimal | time = 1 | 5 sec | 131980 | 21.56 miles | 25.14 minutes | 8.11kWh | 54.81 |
energy-optimal | energy = 1 | 40 sec | 186194 | 19.53 miles | 69 minutes | 2.92 kWh | 92.21 |

the trends all look right here for the distance/time/energy tradeoffs. this is using raw costs so the cost column is kinda nonsense (hence removing the dollar sign).

the a* heuristic is probably still to blame here, given the size of the trees for those searches.

Reply:
yeah it's also curious why the energy value is down to 2.9 (versus 4 something from the previous version).

something else I noticed for the energy optimal case is that the result["cost_summary"]["cost"]["energy_electric"] = 2.92 but the result["cost"] value is 3.19

perhaps we should just lump this into a new issue and do a deep dive after the break?

Reply:
> result["cost_summary"]["cost"]["energy_electric"] = 2.92 but the result["cost"] value is 3.19

right. i think this would be explained if the state updates were allowed zero/negative-valued energies but the cost function was not.

> perhaps we should just lump this into a new issue and do a deep dive after the break?

i at least have an issue for exploring the poor runtimes (#92) - but please feel free to drop another issue on for exploring the energy results as well.